### ISSUE TITLE
Speed and place cells

### ISSUE THREAD

**Issue by hsw28** [RatInABox-Lab/RatInABox#96]
Hi,
I am using ratinabox to model some place cell activity. As we know, place cells are really only active representations of the animals location when the animal is moving/during theta (of course they also activate during SWRs, etc, but those dont represent the current location). I would like to look at the code that "specifies" this property but I cant find it. Would you mind directing me to the right place? Many thanks!!

**Comment by TomGeorge1234** [comment]
Hi Hannah! 

Yes, you're right. Out of the box `PlaceCells` (as they stand) don't model down to that level of biological realism. The goal for RiaB is to provide a minimal set of tools from which this kind of stuff can be explored, so you're on the right path.  

The lines of code which determine the `PlaceCell` firing rates are located in the `get_state()` method of `PlaceCells`. You'll see that, by default, it's just a basic Gaussian and there's no `Agent.velocity` dependence or theta dependence.  
https://github.com/RatInABox-Lab/RatInABox/blob/16375e1d9d654a9b57b7a870aef5588ceb82e627/ratinabox/Neurons.py#L848-L893

So how could you model speed dependency? 

I see two options:
* Subclass `PlaceCells`: You could make a subclass called `MyMoreRealisticPlaceCells(PlaceCells)` which inherits from `PlaceCells` where you write your own `get_state()` method which implements more complex theta or velocity dependence. 
* Mix `PlaceCells` with other cells: You could create a cell class which linearly "mixes" `PlaceCells` with, say, `VelocityCells` using a class we call `FedForwardLayer` to make velocity-dependent-place cells. I did this for `GridCells` in [this demo](https://github.com/RatInABox-Lab/RatInABox/blob/dev/demos/conjunctive_gridcells_example.ipynb)

It depends what you're interested in. The first is probably more flexibility but the second is probably slightly easier. By the way, I once made a subclass of `PlaceCells` which phase precess ([here](https://github.com/RatInABox-Lab/RatInABox/blob/dev/ratinabox/contribs/PhasePrecessingPlaceCells.py) so you might be interested in checking that out too. Once you've decided I'm happy to help you out here if you have specific questions. 

And, of course, once made you'd be welcome to contribute either of these two models to the repo so others can use them too, if you'd like :))

**Comment by hsw28** [comment]
Got it, super helpful and thanks for the fast response! I will likely work on creating a class that makes place cell activity contingent on velocity-- I'll likely have some questions so would you prefer I leave this issue open for those or close it and open a new issue/discussion if need be? Thanks again!

**Comment by TomGeorge1234** [comment]
Np at all. I'll leave it open for now and wait for questions

**Comment by TomGeorge1234** [comment]
not a stupid question! 

* Firstly (and probably the answer) is that the function is called `plot_place_cell_locations()` (not `...._centre`). It's here: 
https://github.com/RatInABox-Lab/RatInABox/blob/e5cc5e91833e0d2cb7a534911fedc8eb1df6d2b9/ratinabox/Neurons.py#L895

* Secondly is that your `CombinedNeurons` subclass should be a subclass of `PlaceCells` (where this function lives) not (or at least not only) `Neurons` 

* Thirdly, unless you want to add extra things into these function you shouldn't need to write such basic wrappers like this. The subclasses "inherit" all methods from their parents by default so `MySubClassInstance.a_method_defined_in_the_parent_class()` should work just fine without you duplicating each function like this. 

* Fourthly, the search bar at the top of GitHub is a pretty effective way to find things/functions. If not, go to [`Neurons.py`](https://github.com/RatInABox-Lab/RatInABox/blob/main/ratinabox/Neurons.py) and poke around there for a while and you should find stuff. 
<img width="1192" alt="Screenshot 2023-12-06 at 10 35 48" src="https://github.com/RatInABox-Lab/RatInABox/assets/41446693/2ba0fecb-9550-4729-9200-4419885ad74a">


Does this help?

**Comment by hsw28** [comment]
great, thank you! my subclass is a subclass of PlaceCells, not just Neurons (sorry if I was unclear). I added the wrappers because I am calling the plotting functions (eg plot_rate_map) from a different main script.

I (of course) tried to search but was searching on plot_place_cell_centres-- it is described in your list of plotting functions (here: https://github.com/RatInABox-Lab/RatInABox/blob/080b4b895463735f5bc498e5773665d05d252275/demos/list_of_plotting_fuctions.md?plain=1#L108 ) which is where I must have pulled it from :) 

Either way, all working now, so thank you!!

**Comment by hsw28** [comment]
Ok, a few other questions:
Here are some plots of my agents movement (taken from real position data)
<img width="533" alt="image" src="https://github.com/RatInABox-Lab/RatInABox/assets/6922509/bbd6607e-feb7-498a-9ef5-89750981263b">
<img width="541" alt="image" src="https://github.com/RatInABox-Lab/RatInABox/assets/6922509/b2e9e7b2-0761-4f53-b290-bb645f3ae132">

And then here are the modelled place cell 
the centers seem to be relatively uniformly distributed despite that the agent is spending significantly more time on the perimeter of the arena. Also, why is there a stripe at the top with no centers, despite that the animal visited that area?
<img width="558" alt="image" src="https://github.com/RatInABox-Lab/RatInABox/assets/6922509/bc1729e3-6b5e-4c34-b52e-880dc0c1e64f">

thank you again!

**Comment by TomGeorge1234** [comment]
Oh sorry for the misnamed plot function. I've fixed that! 

Yes, same as before, in RiaB `PlaceCells` don't model this level of biological realism (this is a conscious choice to keep the toolkit simple). If you wanted you could definitely write your own code that sets `PlaceCells.place_cell_centres` to be _sampled_ from the motion heat map distribution (or something more complex) but, by default, cell centres are just scattered uniformly across the environment plus a bit of jitter so this aspect is totally expected.

As for the band at the top, I just looked into it and it seems like you have spotted a very subtle bug, thanks! It was essentially a fence-post error do with how positioned are sampled uniformly across the `Environment`. I push a fix in version 1.11.4 which should mend things can you let me know. if you agree, update riab with 

```
pip install -U ratinabox
```

**Comment by hsw28** [comment]
great, thank you! if the place cells are scattered randomly across the environment, and can even be in locations where the animal has not travelled, what function does the agents position then have?

i was basing my modified class off of only creating place cells when the agent is travelling at a certain speed, but if the agents position does not matter for creating place cells, the speed will then also be irrelevant

**Comment by hsw28** [comment]
Yes, got it. I'm dreadfully picky and don't just want the PC density to reflect the density of locations, but I want the actual place cell firing to reflect the agent's current position. Terrible, I know... 

I'll see what I can do. thanks!!

**Comment by TomGeorge1234** [comment]
How're you getting on? I'm thinking of closing this issue if theres no more questions for now but up to you

**Comment by hsw28** [comment]
Doing great! I've built 4 models so far haha. Thank you! Feel free to close :)
